cloud/logstream.go

```diff
@@ -2,32 +2,35 @@ package cloud
 
 import (
 	"context"
-	"sync"
+	"io"
+	"sync/atomic"
 
 	"github.com/earthly/cloud-api/logstream"
 	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
 	"github.com/pkg/errors"
 	"golang.org/x/sync/errgroup"
 )
 
-func (c *client) StreamLogs(ctx context.Context, buildID string, deltasCh chan []*logstream.Delta) error {
+// Deltas is a type for iterating over logstream deltas.
+type Deltas interface {
+	Next(ctx context.Context) ([]*logstream.Delta, error)
+}
+
+func (c *client) StreamLogs(ctx context.Context, buildID string, deltas Deltas) error {
 	streamClient, err := c.logstream.StreamLogs(c.withAuth(ctx), grpc_retry.Disable())
 	if err != nil {
 		return errors.Wrap(err, "failed to create log stream client")
 	}
 	eg, ctx := errgroup.WithContext(ctx)
-	var mu sync.Mutex
-	finished := false
+	var finished atomic.Bool
 	eg.Go(func() error {
 		for {
 			resp, err := streamClient.Recv()
 			if err != nil {
 				return errors.Wrap(err, "failed to read log stream response")
 			}
 			if resp.GetEofAck() {
-				mu.Lock()
-				defer mu.Unlock()
-				if !finished {
+				if finished.Load() {
```

Comment:
Shouldn't this be `!finished.Load()`?
The previous code was checking if that finished was false.
I'd expect finished to be set to true here after we sent an EOF and the server responded with an ack.

Reply:
:nod: adding test, fixing; thanks!